pkg/services/ngalert/store/alert_rule.go

```diff
@@ -231,6 +238,73 @@ func (st DBstore) UpdateAlertRules(ctx context.Context, rules []ngmodels.UpdateR
 	})
 }
 
+// preventIntermediateUniqueConstraintViolations prevents unique constraint violations caused by an intermediate update.
+// The uniqueness constraint for titles within an org+folder is enforced on every update within a transaction
+// instead of on commit (deferred constraint). This means that there could be a set of updates that will throw
+// a unique constraint violation in an intermediate step even though the final state is valid.
+// For example, a chain of updates RuleA -> RuleB -> RuleC could fail if not executed in the correct order, or
+// a swap of titles RuleA <-> RuleB cannot be executed in any order without violating the constraint.
+func (st DBstore) preventIntermediateUniqueConstraintViolations(sess *db.Session, updates []ngmodels.UpdateRule) error {
+	// The exact solution to this is complex and requires determining directed paths and cycles in the update graph,
+	// adding in temporary updates to break cycles, and then executing the updates in reverse topological order.
+	// This is not implemented here. Instead, we choose a simpler solution that works in all cases but might perform
+	// more updates than necessary. This simpler solution makes a determination of whether an intermediate collision
+	// could occur and if so, adds a temporary title on all updated rules to break any cycles and remove the need for
+	// specific ordering.
+
+	// This short circuit detects when no intermediate unique constraint violations are possible.
+	// It may return false positives, but never false negatives.
+	if !newTitlesOverlapExisting(updates) {
+		// No possibility of an intermediate unique constraint violation.
+		return nil
+	}
+
+	for _, update := range updates {
+		r := update.Existing
+		u := uuid.New().String()
+		uniqueTempTitle := r.Title + u
+		if len(uniqueTempTitle) > AlertRuleMaxTitleLength {
+			uniqueTempTitle = r.Title[:AlertRuleMaxTitleLength-len(u)] + uuid.New().String()
+		}
+		if updated, err := sess.ID(r.ID).Cols("title").Update(&ngmodels.AlertRule{Title: uniqueTempTitle, Version: r.Version}); err != nil || updated == 0 {
```

Comment:
I wonder if we can simplify it even more and have an intermediate update of the rule's title to it's UID instead of `Title + uid`, i.e. rule.Title = rule.UID. So, it will guarantee uniqueness (because UID has much stronger constraints) and you will not need to take care of the length limits.

Reply:
For UIDs generated by the system, this sounds good and is unlikely to cause any trouble. However, UIDs can also be defined through provisioning / API. If a user creates a human readable UID that happens to overlap with a title change, this would cause a very strange and hard to understand error.

I could simply set the temporary title to just a new random uid, but figured to code defensively here in case it somehow gets persisted by a future bug. They will at least be able to somewhat identify it by having the old title in there.

Reply:
ah, that makes sense!